### PR TITLE
docs: add contributor docs

### PR DESCRIPTION
**Pull request by latonz**
Closes https://github.com/riok/mapperly/issues/231

### CODE REVIEW

File: docs/docs/98-contributing/01-architecture.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+   It instantiates a `User*Mapping` (eg. `UserDefinedNewInstanceMethodMapping`) for each discovered mapping method and adds it to the queue of mappings to work on.
+4. For each mapping in the queue the `DescriptorBuilder` tries to build it's implementation bodies.
+   This is done by a so called `*MappingBodyBuilder`.
+   A mapping body builder tries to map each property from the source to the target.
+   To do this, it asks the `DescriptorBuilder` to create mappings for the according types.
+   To create a mapping from one type to another, the `DescriptorBuilder` loops through a set of `*MappingBuilder`s.
+   Each of the mapping builders try to create a mapping (an `ITypeMapping` implementation) for the asked type mapping by using
+   one approach on how to map types (eg. an explicit cast is implemented by the `ExplicitCastMappingBuilder`).
+   These mappings are queued in the queue of mappings which need the body to be built (currently body builders are only used for object to object (property-based) mappings).
+5. The `SourceEmitter` emits the code described by the `MapperDescriptor` and all it's mappings.
```
Comment (CommonGuy) [inline docs/docs/98-contributing/01-architecture.md:33]: Same as above -> "its"

File: docs/docs/98-contributing/01-architecture.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Roslyn multi targeting
+
+Mapperly targets multiple roslyn versions by building multiple NuGet packages
```
Comment (CommonGuy) [inline docs/docs/98-contributing/01-architecture.md:37]: Roslyn is spelled with an uppercase R

File: docs/docs/98-contributing/02-tests.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+## Integration tests
+
+The integration tests are located in `tests/Riok.Mapperly.IntegrationTests`.
+Integration tests are "a bit of everything" mappers.
+These tests run locally by referencing the source generator as an analyzer.
+In the CI pipeline, the integration tests reference the built NuGet package and
+are run on several supported target frameworks (including .NET 7.0 but also .NET Framework).
+
+Debugging integration tests is a lot harder than debugging unit tests.
+Therefore if a integration test needs to be debugged,
```
Comment (CommonGuy) [inline docs/docs/98-contributing/02-tests.md:29]: "an" integration test

File: docs/docs/98-contributing/02-tests.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+it is often easier to implement an unit test for the to be tested behaviour
+and debug the unit test instead of the integration test.
+See also the [debugging](./debugging) documentation topic on how to debug Mapperly.
```
Comment (CommonGuy) [inline docs/docs/98-contributing/02-tests.md:32]: Would include "documentation" in the link (currently it is outside the link)

File: docs/docs/98-contributing/03-debugging.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Debugging
+
+To debug the code generated by Mapperly just set a breakpoint in the generated code and you are happy to go.
```
Comment (CommonGuy) [inline docs/docs/98-contributing/03-debugging.md:3]: Is "happy to go" a common expression? I only know about "ready to go"

File: docs/docs/98-contributing/01-architecture.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+This describes the process implemented by Mapperly on a higher level.
+For each discovered `MapperAttribute` a new `DescriptorBuilder` is created.
+The `DescriptorBuilder` is responsible to build a `MapperDescriptor` which holds all the mappings.
+The `DescriptorBuilder` does this by following this process:
+1. Extracting the configuration from the attribute
+2. Extracting user implemented object factories
+3. Extracting user implemented and user defined mapping methods.
+   It instantiates a `User*Mapping` (eg. `UserDefinedNewInstanceMethodMapping`) for each discovered mapping method and adds it to the queue of mappings to work on.
+4. For each mapping in the queue the `DescriptorBuilder` tries to build it's implementation bodies.
```
[the remark this answers is not in the document]
Comment (latonz) [reply docs/docs/98-contributing/01-architecture.md]: Fixed

File: docs/docs/98-contributing/01-architecture.md (hunk 1 above)
Comment (latonz) [reply docs/docs/98-contributing/01-architecture.md]: Fixed

File: docs/docs/98-contributing/01-architecture.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+   To create a mapping from one type to another, the `DescriptorBuilder` loops through a set of `*MappingBuilder`s.
+   Each of the mapping builders try to create a mapping (an `ITypeMapping` implementation) for the asked type mapping by using
+   one approach on how to map types (eg. an explicit cast is implemented by the `ExplicitCastMappingBuilder`).
+   These mappings are queued in the queue of mappings which need the body to be built (currently body builders are only used for object to object (property-based) mappings).
+5. The `SourceEmitter` emits the code described by the `MapperDescriptor` and all it's mappings.
+
+## Roslyn multi targeting
+
+Mapperly targets multiple roslyn versions by building multiple NuGet packages
+and merging them together into a single one. See `build/package.sh` for details.
```
[the remark this answers is not in the document]
Comment (latonz) [reply docs/docs/98-contributing/01-architecture.md]: Fixed

File: docs/docs/98-contributing/01-architecture.md (hunk 2 above)
Comment (latonz) [reply docs/docs/98-contributing/01-architecture.md]: Fixed

File: docs/docs/98-contributing/02-tests.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+You can run the tests by running
+```bash
+dotnet test
+```
+or by running the discovered tests in your IDE.
+
+## Unit tests
+
+The unit tests are located in the `tests/Riok.Mapperly.*.Tests` projects.
+These unit tests usually test exactly one unit of code or one type of mapping by isolating it's code
```
[the remark this answers is not in the document]
Comment (latonz) [reply docs/docs/98-contributing/02-tests.md]: Fixed

File: docs/docs/98-contributing/02-tests.md (hunk 3 above)
Comment (latonz) [reply docs/docs/98-contributing/02-tests.md]: Fixed

File: docs/docs/98-contributing/02-tests.md (hunk 4 above)
Comment (latonz) [reply docs/docs/98-contributing/02-tests.md]: Improved

File: docs/docs/98-contributing/03-debugging.md (hunk 5 above)
Comment (latonz) [reply docs/docs/98-contributing/03-debugging.md]: replaced with ready to go

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**